funding/manager.go

```diff
@@ -1192,8 +1203,8 @@ func (f *Manager) advancePendingChannelState(
 	}
 
 	confChannel, err := f.waitForFundingWithTimeout(channel)
-	if err == ErrConfirmationTimeout {
-		return f.fundingTimeout(channel, pendingChanID)
+	if err == ErrConfirmationTimeout || err == errFundingInputSpent {
+		return f.fundingTimeout(channel, pendingChanID, err)
```

Comment:
then the channel is lost. this is an issue with `RegisterSpendNtfn` only waiting for 1 confirmation before notifying. i don't think adding the logic is worth it, the better fix is to fix the api

Reply:
And even if the "channel is lost", are we able to force close it rather than losing funds?

Reply:
this is pre-existing, the chain-watching code in the contractcourt will make you lose funds if there is a reorg and a conflicting close tx confirms since it uses `RegisterSpendNtfn`

Reply:
Even with the existing API, it seems unnecessary to lose funds in this case.  Looks like `SpendEvent` contains a `Reorg` and a `Done` channel specifically for this scenario.

So we can immediately unlock our UTXOs on a spend notification and completely forget about the pending channel once `Done` is sent upon.

For contract court, we should really be doing something similar.

Reply:
I don't think this belongs in this PR - making lnd reorg-safe should probably be a larger change in a separate PR. This issue isn't isolated to the contractcourt either, if the counter-party sets `MinAcceptDepth` to 1 in accept_channel, we'll consider the channel open after a single-conf. If a reorg then happens (more likely if depth is only 1), we'll have the wrong SCID and the channel won't be usable by the broader network. 

The `Done` channel is sent on after the mature block height is reached, which is 144 blocks after spend/conf - so it's unreasonable to wait that long. The issue with the `Reorg` channel is that it requires the callsite to add its own logic surrounding it (i.e. counting the number of blocks before the coast is clear), since all of lnd's callsites will:
- call `RegisterSpendNtfn`
- get notified after a single confirmation containing the spend and proceed
- not listen on the `Reorg` channel

I am leaning towards not adding any logic here in this PR, but others can chime in. I would prefer that the reorg change is uniform across the board and not a one-off we have to remove later. I am fine with this PR lingering until a change like that is in

Also, the contractcourt logic doesn't seem to have an easy fix either, since lnd will go to chain 10 blocks before the timeout of an incoming HTLC, and if lnd has a 6-block reorg buffer before considering the channel closed, the window is now 4 blocks. We could increase the 10 blocks to something higher, but looking on 1ml, some users have CLTV deltas set to 18 and increasing the hard-coded 10 blocks could lead to unnecessary force closes

Reply:
I think every case described here is properly addressed using the "act immediately, forget important stuff after 144 confirms" pattern from the existing API.  The "wait for 6 confs, then act and forget stuff" pattern will be simpler in some cases, but it doesn't work well for everything (as you mention for contract court) and introduces tail risk of losing funds -- what if we have a black swan reorg like the 20+ block ones in the past?

> This issue isn't isolated to the contractcourt either, if the counter-party sets `MinAcceptDepth` to 1 in accept_channel, we'll consider the channel open after a single-conf. If a reorg then happens (more likely if depth is only 1), we'll have the wrong SCID and the channel won't be usable by the broader network.

Yep.  But at least today we can close the channel and not lose funds. (Right?)

> The `Done` channel is sent on after the mature block height is reached, which is 144 blocks after spend/conf - so it's unreasonable to wait that long.

I don't think it's terribly unreasonable for cases where we lose funds on reorg.  For this PR, we'd just be storing some extra bytes for a day, so we at least know how to recover funds if the funding tx is reorg'd in.

> Also, the contractcourt logic doesn't seem to have an easy fix either, since lnd will go to chain 10 blocks before the timeout of an incoming HTLC, and if lnd has a 6-block reorg buffer before considering the channel closed, the window is now 4 blocks. We could increase the 10 blocks to something higher, but looking on 1ml, some users have CLTV deltas set to 18 and increasing the hard-coded 10 blocks could lead to unnecessary force closes

Isn't this exactly the reason we should act immediately upon the first confirmation, but not forget the things until X confirmations?

For the incoming HTLC example, we broadcast the commitment tx and immediately intitate htlc-success flow, but don't forget about prior commitment txs or revocation keys until 144 blocks.

Or another example:  peer broadcasts a revoked commitment, we punish immediately on first confirmation but don't forget other commitments or revocation keys until 144 blocks.

> The issue with the `Reorg` channel is that it requires the callsite to add its own logic surrounding it (i.e. counting the number of blocks before the coast is clear)

Intended use seems to be a select on `Reorg` and `Done` channels.  No need to count blocks ourselves if we're OK with 144 blocks before forgetting important things.



I might be missing something, but it seems in all these cases we can prevent fund loss with the existing API, and the suggested new API might make 1 of the cases simpler to handle (double-spend of funding tx input).

Reply:
Waiting for 144 blocks is a large change to the contractcourt. Basically everything would need to be rewound. Currently once an event happens, it's forgotten and state is eventually deleted.

Reply:
This specific discussion is related but not about the channel confirmations. We could enforce 6 confs, but also know that if you open a channel to somebody else, they choose the min confs and a reorg could make your channel unusable (but not cause funds loss). Also the spec doesn't really prescribe what to do if your SCID changes. I'll change this code to register for confirmations, since it seems pretty straightforward. But changing the contractcourt is probably a bit of a lift and isn't so straightforward...

Reply:
Also related is this OG PR https://github.com/lightningnetwork/lnd/issues/53 that might be a bit more enlightening

Reply:
I've decided not to change the code here. If you want to make this and other code re-org safe, feel free. Already you can lose funds due to a funding transaction getting reorg'd out and replaced with a double-spend, so I don't think this patch makes things much worse

Reply:
> Already you can lose funds due to a funding transaction getting reorg'd out and replaced with a double-spend, so I don't think this patch makes things much worse

I don't understand this logic -- if we double-spend our own UTXO in a reorg, we should strictly benefit (able to spend the same UTXO both on and off chain).  If the peer is initiator and could double-spend, that's our fault for choosing too small of a `minimum_depth`.

This patch does make things worse.  Before this patch, we never forget the funding tx, so we can always recover if the funding tx is reorg'd in.  After this patch, we cannot.

Reply:
>  Before this patch, we never forget the funding tx, so we can always recover if the funding tx is reorg'd in. After this patch, we cannot.

The current issue with LND is that if there is a double spend it'll be stuck in funding flow never being able to initiate funding with the same peer again. There are various legit reasons to double spend so "whatever, the peer was an asshole anyway" isn't sufficient.

Reply:
@morehouse Can you review the latest diff here? It should address the reorg situation. I chose not to use `Done` since it is not always sent on if the confirmation/spend notification is old. Instead of changing the txnotifier logic in this PR, I made the funding code just register for a confirmation notification of 144 blocks. If the conflict tx doesn't get 144 confirmations in a period of 160 blocks, we'll fall back to the old behavior in case the funding tx actually gets reorg'd back in. With the changes, I don't _think_ anything should be worse now

Reply:
The reorg approach LGTM, though it's a shame we can't use `Done`, as it would simplify the funding manager side.

Still need a test for funding tx being reorged in.